Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,520 @@
+       # This would be a bad random number generator!
+       print("Hi")
+       yield 1
+       print("Hello")
+       yield 7
+       print("Howdy")
+       yield 4
+       ...
+
+Similar to a coroutine function, calling a generator function does not run it.
+Instead, it creates a generator object::
+
+   >>> get_random_number()
+   <generator object get_random_number at 0x1048671c0>
+
+You can proceed to the next ``yield`` of a generator by using the
+built-in function :func:`next`.
+In other words, the generator runs, then pauses.
+For example::
+
+   >>> generator = get_random_number()
+   >>> next(generator)
+   Hi
+   1
+   >>> next(generator)
+   Hello
+   7
+
+=====
+Tasks
+=====
+
+Roughly speaking, :ref:`tasks <asyncio-task-obj>` are coroutines (not coroutine
+functions) tied to an event loop.
+A task also maintains a list of callback functions whose importance will become
+clear in a moment when we discuss :keyword:`await`.
+The recommended way to create tasks is via :func:`asyncio.create_task`.
+Creating a task automatically adds it to the event loop's queue of tasks.
+
+Since there's only one event loop (in each thread), :mod:`!asyncio` takes care of
+associating the task with the event loop for you. As such, there's no need
+to specify the event loop.
+
+::
+
+   coroutine = loudmouth_penguin(magic_number=5)
+   # This creates a Task object and puts it on the event loop's queue.
+   task = asyncio.create_task(coroutine)
+
+Earlier, we manually created the event loop and set it to run forever.
+In practice, it's recommended to use (and common to see) :func:`asyncio.run`,
+which takes care of managing the event loop and ensuring the provided
+coroutine finishes before advancing.
+For example, many async programs follow this setup::
+
+   import asyncio
+
+   async def main():
+       ...
+
+   if __name__ == "__main__":
+       asyncio.run(main())
+       # The program will not reach the following print statement until the
+       # coroutine main() finishes.
+       print("coroutine main() is done!")
+
+=====
+await
+=====
+
+:keyword:`await` is a Python keyword that's commonly used in one of two
+different ways::
+
+   await task
+   await coroutine
+
+Unfortunately, it does matter which type of object is ``await``\ ed.
+
+``await``\ ing a task will cede control from the current task or coroutine to
+the event loop.
+In the process of relinquishing control, the task that's giving up control
+adds a callback to the ``await``\ ed task's list of callbacks indicating it
+should resume the current task/coroutine when it (the ``await``\ ed one)
+finishes.
+In other words, when that ``await``\ ed task finishes, the original task is
+added back to the event loops queue.
+
+This is a basic, yet reliable mental model.
+In practice, it's slightly more complex, but not by much.
+In part 2, we'll walk through the details that make this possible.
+
+**Unlike tasks, awaiting a coroutine does not hand control back to the event
```

Comment:
This isn't entirely correct, using tasks does not guarantee that the control will be given to event loop if e.g. the task completes eagerly.

Reply:
Thanks for the clarification. I didn't know about that! You're referring to the `eager_start=True` parameter in 
`create_task()`, right?

We could add a brief note, like: "Unlike tasks, awaiting a coroutine does not hand control back to the event loop (though note that tasks created with eager_start=True may also complete without yielding control)."

Though, since this parameter defaults to `False` and the documentation for `eager_start` pretty well already describes this special behavior I lean towards leaving this as-is. I'm generally wary of documenting the various edge cases that people can opt-in to which the style guide also somewhat advises against: https://devguide.python.org/documentation/style-guide/#economy-of-expression